maven/JarAssembler.kt

```diff
@@ -34,35 +36,73 @@ class JarAssembler : Callable<Unit> {
     @Option(names = ["--jars"], split = ";")
     lateinit var jars: Array<File>
 
-    val entryNames = mutableSetOf<String>()
+    private val entryNames = mutableSetOf<String>()
+    val entries = HashMap<String, ByteArray>()
+
+    /**
+     * For path "a/b/c.java" inserts "a/" and "a/b/ into `entries`
+     */
+    private fun preCreateDirectories(path: Path) {
+        for (i in path.nameCount-1 downTo 1) {
+            val subPath = path.subpath(0, i).toString() + "/"
+            entries[subPath] = ByteArray(0)
+        }
+    }
+
+    private fun getFinalPath(entry: ZipEntry, sourceFileBytes: ByteArray): String {
+        return if (entry.name.endsWith(".java")) {
+            // files in source JARs are moved according to their `package` statement
+            val sourceFile = sourceFileBytes.toString(Charset.forName("UTF-8"))
+            val sourceFileName = Paths.get(entry.name).fileName.toString()
+            val sourceFilePackage = javaPackageRegex.find(sourceFile)?.groups?.get(1)?.value ?: throw RuntimeException("could not obtain package of ${entry.name}")
+            "${sourceFilePackage.replace(".", "/")}/$sourceFileName"
+        } else {
+            entry.name
+        }
+    }
+
+    private fun prepareEntry(inputStream: BufferedInputStream, entry: ZipEntry) {
+        val sourceFileBytes = inputStream.readBytes()
+        val resultLocation = getFinalPath(entry, sourceFileBytes)
+        entries[resultLocation] = sourceFileBytes
+        preCreateDirectories(Paths.get(resultLocation))
+    }
 
     override fun call() {
         ZipOutputStream(BufferedOutputStream(FileOutputStream(output_file))).use { out ->
             if (pomFile != null) {
-                val pomFileEntry = ZipEntry("META-INF/maven/${groupId}/${artifactId}/pom.xml")
-                out.putNextEntry(pomFileEntry)
-                out.write(pomFile!!.readBytes())
+                val pomPath = "META-INF/maven/${groupId}/${artifactId}/pom.xml"
+                entries[pomPath] = pomFile!!.readBytes()
```

Comment:
Can we make `preCreateDirectories` _side-effect-free_ and returns created entries? Then, you would merge those entries with the global `entries` variable:

```
                entries += preCreateDirectories(Paths.get(pomPath))
                entries[pomPath] = pomFile!!.readBytes()
```

Reply:
Also I think it's nicer if you do the pre-create before rather than after.

Reply:
@lolski addressed in https://github.com/vaticle/bazel-distribution/pull/305/commits/cad87304cd7785647fcfc952b1fd3152cf396490